include/lbann/layers/transform/upsample.hpp

```diff
@@ -0,0 +1,297 @@
+      m_scale_factors(other.m_scale_factors)
+#ifdef LBANN_HAS_DNN_LIB
+      ,
+      m_pooling_dnn_desc(other.m_pooling_dnn_desc),
+      m_tensors_dnn_desc(other.m_tensors_dnn_desc)
+#endif // LBANN_HAS_DNN_LIB
+  {
+#ifdef LBANN_HAS_DNN_LIB
+    m_tensors_dnn_desc.set_layer(this);
+#endif // LBANN_HAS_DNN_LIB
+  }
+
+  upsample_layer& operator=(const upsample_layer& other)
+  {
+    data_type_layer<TensorDataType>::operator=(other);
+    m_upsample_mode = other.m_upsample_mode;
+    m_scale_factors = other.m_scale_factors;
+#ifdef LBANN_HAS_DNN_LIB
+    m_pooling_dnn_desc = other.m_pooling_dnn_desc;
+    m_tensors_dnn_desc = other.m_tensors_dnn_desc;
+    m_tensors_dnn_desc.set_layer(this);
+#endif // LBANN_HAS_DNN_LIB
+    return *this;
+  }
+
+  ~upsample_layer() override = default;
+
+  upsample_layer* copy() const override { return new upsample_layer(*this); }
+
+  /** @name Serialization */
+  ///@{
+
+  template <typename ArchiveT>
+  void serialize(ArchiveT& ar);
+
+  ///@}
+
+  std::string get_type() const override { return "upsample"; }
+  data_layout get_data_layout() const override { return T_layout; }
+  El::Device get_device_allocation() const override { return Dev; }
+
+#ifdef LBANN_HAS_ONNX
+  void fill_onnx_node(onnx::GraphProto& graph) const override;
+#endif // LBANN_HAS_ONNX
+
+  description get_description() const override
+  {
+    auto desc = data_type_layer<TensorDataType>::get_description();
+    std::stringstream ss;
+
+    // Upsample mode
+    ss.str(std::string{});
+    ss.clear();
+    switch (m_upsample_mode) {
+    case upsample_mode::NEAREST:
+      ss << "nearest";
+      break;
+    default:
+      ss << "invalid";
+    }
+    desc.add("Upsample mode", ss.str());
+
+    // Upsample scale factors
+    ss.str(std::string{});
+    ss.clear();
+    for (size_t i = 0; i < m_scale_factors.size(); ++i) {
+      ss << (i > 0 ? ", " : "") << m_scale_factors[i];
+    }
+    desc.add("Scale factors", ss.str());
+
+    // Result
+    return desc;
+  }
+
+protected:
+  /** Add layer specific data to prototext */
+  void write_specific_proto(lbann_data::Layer& proto) const final;
+
+  friend class cereal::access;
+  upsample_layer() : upsample_layer(nullptr, 1, 1, upsample_mode::NEAREST) {}
+
+  void setup_dims() override
+  {
+    data_type_layer<TensorDataType>::setup_dims();
+    const auto& input_dims = this->get_input_dims();
+    auto output_dims = input_dims;
+    for (size_t i = 0; i < output_dims.size() - 1; ++i) {
+      output_dims[i + 1] = m_scale_factors[i] * input_dims[i + 1];
+    }
+    this->set_output_dims(output_dims);
+  }
+
+  /// Initialize GPU objects
+  void setup_gpu() override
+  {
+    data_type_layer<TensorDataType>::setup_gpu();
+#ifndef LBANN_HAS_DNN_LIB
+    LBANN_ERROR("DNN library not detected");
+#else
+
+    // Set upsample descriptor
+    int ndims = m_scale_factors.size();
+    std::vector<int> padding(ndims, 0);
+    m_pooling_dnn_desc.set(pooling_mode::AVERAGE_COUNT_INCLUDE_PADDING,
```

Comment:
Is it then "nearest" rather than linear interpolation?

Reply:
No, I am reusing mean pooling backwards for the forward pass as it behaves as a nearest neighbor upsampling with a scale factor.